snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,309 @@
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
+
+
+@dataclass
+class Subscription(Generic[TStream, TOffset]):
+    topics: Sequence[Topic[TStream]]
+    remote_consumer_group_offsets: MutableMapping[
+        TStream, MutableMapping[ConsumerGroup, TOffset]
+    ] = field(default_factory=lambda: defaultdict(dict))
+
+    def get_effective_remote_consumer_group_offset(
+        self, stream: TStream, groups: Iterable[ConsumerGroup]
+    ) -> Optional[TOffset]:
+        offsets = self.remote_consumer_group_offsets[stream]
+        return min(
+            filter(
+                lambda offset: offset is not None,
+                [offsets.get(group) for group in groups],
+            ),
+            default=None,
+        )
+
+
+class SynchronizedConsumerBackend(ConsumerBackend[TStream, TOffset, TValue]):
+    def __init__(
+        self,
+        backend: ConsumerBackend[TStream, TOffset, TValue],
+        commit_log_consumer_backend: ConsumerBackend[
+            # NOTE: These TStream and TOffset values for the commit log
+            # consumer do not need to match, but this simplifies things a bit
+            # so that we don't have to declare two additional type variables.
+            TStream,
+            TOffset,
+            CommitData[TStream, TOffset],
+        ],
+        remote_consumer_groups: Set[ConsumerGroup],
+        get_commit_log_streams_for_topic: Callable[[Topic[TStream]], Sequence[TStream]],
+    ) -> None:
+        self.__backend = backend
+        self.__commit_log_consumer_backend = commit_log_consumer_backend
+        self.__remote_consumer_groups = remote_consumer_groups
+        self.__get_commit_log_streams_for_topic = get_commit_log_streams_for_topic
+
+        self.__subscription: Synchronized[
+            Subscription[TStream, TOffset]
+        ] = Synchronized(Subscription([]))
+
+        self.__paused_streams: Set[TStream] = set()
+
+        self.__shutdown_requested = Event()
+
+        # TODO: If the commit log consumer crashes, it should cause all method
+        # calls on this object to raise (probably a RuntimeError raised from
+        # the commit log consumer exception.)
+        self.__commit_log_consumer_future = execute(self.__run_commit_log_consumer)
+
+    def __run_commit_log_consumer(self) -> None:
+        # TODO: How do we ensure that the commit log consumer is not committing
+        # it's offsets (or is subject to automatic offset resets?)
+
+        # XXX: This reference to the subscription object is only maintained so
+        # that it can be referenced checked during the commit log consumer main
+        # loop. It should not be used otherwise!
+        current_subscription: Optional[Subscription[TStream, TOffset]] = None
+
+        while not self.__shutdown_requested.is_set():
+            with self.__subscription.get() as subscription:
```

Comment:
The subscription object is referenced in two different threads — this one (the commit log consumer run loop) and the "local" consumer run loop thread (the main thread if you were to instantiate and run the `SynchronizedConsumer` there.)

Since this is acting on a collection rather than scalars, it's not safe to trust the GIL either for mutual exclusion here since mutations or reads to multiple key/value pairs may be interleaved (though I'd argue the "trust the GIL" is a bad philosophy regardless.)